### ISSUE TITLE
promotion of float with int never increases precision of float

### ISSUE THREAD

**Issue by nairbv** [pytorch/pytorch#34167]
I noticed that:

```
In [4]: (torch.ones(5, dtype=torch.float) + torch.ones(5, dtype=torch.long)).dtype
Out[4]: torch.float32

# where as in numpy:
In [5]: (np.ones(5, dtype=np.float) + np.ones(5, dtype=np.int64)).dtype
Out[5]: dtype('float64')
```

I would expect the numpy behavior.

This results from the values in the _promoteTypesLookup table in ScalarType.h. The values for floating types don't change regardless of the size of the integral type they're being combined with.

@gchanan do you agree that the numpy would be better here? (the change I'm suggesting wouldn't change behavior for `tensor + scalar` type ops).

**Comment by gchanan** [comment]
No, I don't agree.

We probably captured the thinking in some of the initial type promotion PRs but roughly:
1) people very rarely train in fp64, as it is way slower and doesn't typically improve accuracy.  Most of the pull is to use precisions lower than fp32 (bfloat16, quantization).  If you really want fp64, you can set the default dtype and you'd get the promotion behavior for free.
2) Our APIs are likely to give you longs instead of reduced integer ranges, i.e. `torch.tensor([5])` gives you a long tensor and all our index APIs use long tensors.  Because reading values on GPU requires a sync point, it's also not the best idea to make the behavior dependent on the stored values, either.

Thus, you don't really have flexibility on the floating-point end or on the integer end.

**Comment by nairbv** [comment]
I don't think we modified the ScalarType.h matrix of types in any of the initial PRs. We did add logic to ignore the dtype of the int if there was a difference in kind/dimension, but looking back it feels like some of that specificity wasn't necessary if we actually don't want to promote floats when the dimension is the same.

> If you really want fp64, you can set the default dtype and you'd get the promotion behavior for free.

that doesn't affect this case though. Even if using half + long, the result is `half`, not `get_default_dtype`. If that's desired that's fine, but it seems odd to me to go this far with it:

```
> (torch.ones(5, dtype=torch.half, device='cuda') + torch.ones(5, dtype=torch.long, device='cuda')).dtype
torch.float16
```
If we wanted a float32 result I think we'd have to manually convert the half tensor first.

> it's also not the best idea to make the behavior dependent on the stored values

right, that was a very conscious decision. I guess if it's common to combine float types with integral values coming from indices etc, then we shouldn't want to consider the integral types.

The reason I'm thinking about this is that the "categories" of:
`floating > integral > boolean` 
will become 
`complex > floating > integral > boolean`
and I want to make sure we can clearly explain the resulting behavior in a consistent way.

Right now we have `torch.promote_types(torch.half, torch.long) => torch.half` (not promoting), but `torch.promote_types(torch.complex64, torch.double) => torch.complex128` (promoting) and that seems sensible/reasonable but still slightly unexpected given our category/kind docs.

**Comment by gchanan** [comment]
> that doesn't affect this case though. Even if using half + long, the result is half, not get_default_dtype. If that's desired that's fine, but it seems odd to me to go this far with it:

Right, but I don't think that's a common pattern, at least with the new APIs.  The common pattern is to just not specify a dtype or to use something like Apex to handle it for you.

> Right now we have torch.promote_types(torch.half, torch.long) => torch.half (not promoting), but torch.promote_types(torch.complex64, torch.double) => torch.complex128 (promoting) and that seems sensible/reasonable but still slightly unexpected given our category/kind docs.

I agree this seems a little strange but I don't think it's really a comparable situation; a `torch.double` can be perfectly represented as a `torch.complex128`; it "is a" `complex128`.  A `torch.int64` isn't a `torch.double`, and can't be perfectly represented as a `torch.double` anyway.  And again, if we take the view that you "opt-in" to using `torch.double` (because none of the APIs give you doubles unless you ask), then it seems fine to promote in that way because you've already opted-in at that point.
